File: MASICPeakFinder/clsSmoothedYDataSubset.vb
Imports System.Collections.Generic

Public Class clsSmoothedYDataSubset
    Public ReadOnly Property DataCount As Integer
    Public ReadOnly Property Data As Single()
    Public ReadOnly Property DataStartIndex As Integer

    Public Sub New()
        DataCount = 0
        DataStartIndex = 0
        ReDim Data(0)
    End Sub

    Public Sub New(yData As IList(Of Double), startIndex As Integer, endIndex As Integer)

        If yData Is Nothing OrElse endIndex < startIndex OrElse startIndex < 0 Then
            DataCount = 0
            DataStartIndex = 0
            ReDim Data(0)
            Return
        End If

        DataStartIndex = startIndex

        DataCount = endIndex - startIndex + 1
        ReDim Data(DataCount)

        For intIndex = startIndex To endIndex
            Data(intIndex - startIndex) = CSng(Math.Min(yData(intIndex), Single.MaxValue))
        Next

    End Sub
End Class

File: MASICPeakFinder/clsSmoothedYDataSubset.cs
using System;
using System.Collections.Generic;

namespace MASICPeakFinder
{
    public class clsSmoothedYDataSubset
    {
        public int DataCount { get; private set; }
        public double[] Data { get; private set; }
        public int DataStartIndex { get; private set; }

        public clsSmoothedYDataSubset()
        {
            DataCount = 0;
            DataStartIndex = 0;
            Data = new double[1];
        }

        public clsSmoothedYDataSubset(IList<double> yData, int startIndex, int endIndex)
        {
            if (yData == null || endIndex < startIndex || startIndex < 0)
            {
                DataCount = 0;
                DataStartIndex = 0;
                Data = new double[1];
                return;
            }

            DataStartIndex = startIndex;

            DataCount = endIndex - startIndex + 1;
            Data = new double[DataCount + 1];

            for (var intIndex = startIndex; intIndex <= endIndex; intIndex++)
                Data[intIndex - startIndex] = Math.Min(yData[intIndex], double.MaxValue);
        }
    }
}